EDDI simple internationalization

fbrossar25 opened the pull request:
Hello,
I programmed a simple internationalization class that read strings in resources files contained in Utilities.lang.
I wanted to share it with you in the hope of earing EDDI talk to me in french :p
The class is Utilities.I18N and I wrote some tests in Tests.I18NTests.
Anyway, thanks for your work.

PS: tell me if you want me to help you more with this (like annoying I18N calls ^^)

# TODO list
- [x] EDDI tab
- [x] Help window
- [x] TTS tab
- [x] Frontier API tab
- [x] Commander Details tab
- [x] EDDP Monitor tab
- [x] Galnet Monitor tab
- [x] Material Monitor tab
- [x] Materials list (without write into the original)
- [x] Ship Monitor tab
- [x] Phonetics
- [x] EDSM Responder tab
- [x] EDDN Responder tab
- [x] EDDP Responder tab
- [x] Speech Responder tab
- [x] Speech Responder's scripts
- [x] VoiceAttack Responder tab
- [x] Roles

Tkael commented:
This is an interesting one. I have a couple of questions / observations:
1. Utilities/lang/en.Designer.cs appears to be at least partially in French?
2. Why use xml for storing language configurations? Wouldn't json be easier to read / edit? A json file could easily contain values for various languages in a single file.
3. As I'm sure you are aware, we've made some changes to the UI with recent versions. Is this up to date with any recent UI changes?

fbrossar25 commented:
Well you've got some good observations, as I'm directly editing with Visual Studio, I didn't think about JSON but I'm sure I can change that easily, and that explain also all those comments that was generated in my language sorry for that ^^.
I'm not sure which UI changes you're talking about but I saw that there are some text in .xaml file that I don't know how to translate them, I mean with my I18N class. All others texts are just hard coded into .cs file which can be updated.

fbrossar25 commented:
OK, It took me some time to deal with JSON but it's finally working well and as you said editing will be easier now ;)

Tkael commented:
Nice. Regarding the xaml, you may wish to investigate some of the examples in the current code for how this can be changed automatically. For example, [https://github.com/EDCD/EDDI/blob/3574b51489802a7a9a5afb3d26e98cf25e9725e0/EDDI/MainWindow.xaml#L21](here) and [https://github.com/EDCD/EDDI/blob/3574b51489802a7a9a5afb3d26e98cf25e9725e0/EDDI/MainWindow.xaml.cs#L288](here).

fbrossar25 commented:
thanks for those examples I'll do that as well as a menu to manually choose the language once i've got some time ^^

fbrossar25 commented:
Not really related but while playing around Ceos-Sothis, the SpeechService was pausing almost a second while saying 'Federation Unite!'. I think it's the same with all factions with any punct char in their name.

richardbuckle commented:
This is great! I am very much in favour of internationalisation.

But I would like to gain a fuller understanding of the best practices for .NET development before committing to any particular approach. I am hesistant because I don't want to go down the wrong path.

If anyone who is more familiar with the .NET way of doing things can educate me and the rest of the team, I'd be extremely grateful.

fbrossar25 commented:
Well for this I can't really teach you anything since I'm not a C# nor a internationalisation expert, so if you find or learn some tips don't forget me ;)
For now i'll continue with this way but I think it's better to wait for a full translation of all texts before comitting anything into the project.

Tkael commented:
Ping ECDC for feedback?

fbrossar25 commented:
Yup, i'll let know when it's done by posting there if you don't mind.

Syrrianyl commented:
It's good than some else work on it, because I'm always read about how to do while you make that.
Just one question : is it possible to have a choice (like combobox) for (french or other) people who play in english and prefer to have EDDi in english also?

fbrossar25 commented:
Well I've already thought about a combobox to do this I just don't have much time during the week with my studies ^^

fbrossar25 commented:
I've added a combobox which let you choose the language you want in EDDI tab. When you chang this setting, EDDI will restart automatically. Also i've added a property in EDDI's configuration in order to save the last language used. Tell me what you think of it ^^. For today I think i'll stop there for the overall translation.

fbrossar25 commented:
Thanks to @Syrrianyl for his work on all the phonetics that he sent to me ;)

fbrossar25 commented:
# TODO list
- [x] EDDI tab
- [x] Help window
- [x] TTS tab
- [x] Frontier API tab
- [x] Commander Details tab
- [ ] EDDP Monitor tab
- [ ] Galnet Monitor tab
- [ ] Material Monitor tab
- [ ] Materials list (without write into the original)
- [ ] Ship Monitor tab
- [x] Phonetics
- [ ] EDSM Responder tab
- [ ] EDDN Responder tab
- [ ] EDDP Responder tab
- [ ] Speech Responder tab
- [ ] Speech Responder's scripts
- [ ] VoiceAttack Responder tab

Tkael commented:
We may wish to be careful about translating material and commodity definitions. The English names are used to match data from the Frontier API.

fbrossar25 commented:
Yup i've seen that I think I will just translate all materials names in my json and call I18N.GetString with the english name in your json to avoid problems

Tkael commented:
Do the same with modules too I think

richardbuckle commented:
This is good work, but I am wondering if a pull request is still appropriate given that work is still going on.

Maybe we need to confer so as to co-ordinate the work flow, and maybe segregate the compile-time changes (XAML and C#) from the data resource changes (JSON etc)?

Thoughts welcomed.

fbrossar25 commented:
It can be a good idea to separate compile-time and datas changes, but I don't understand clearly what you want me to do ^^

Tkael commented:
compile-time changes = the basic structural changes to allow EDDI to support internationalization / translations.
data resource changes = the actual translations

fbrossar25 commented:
Well the problem is that we need to call I18N.GetString pretty much everywhere (compile time) but we also need an existing translation in langs.json (data resource) for each calls otherwise it displays/says nothing. What I can do if you prefer is close that pull request and reopen another one when the work is finished.

StivlWrith commented:
@Tkael for the data that we can found in DataDefinitions, we don't change .nane, we just add a .LocalName to not disturb inside the software
For exemple in the Crime.cs, there is 2 string:
- edname
- name

we add a string LocalNane to have the one localised. ( I use caps to be clear thart it is not the same thing when you change a script)

fbrossar25 commented:
Hi !
This time I think it's finally done. The only thing left is scripts that need to be translated but that can done with an additional personnality.

hoksilato2 commented:
What can I do for translate EDDI to spanish?

StivlWrith commented:
In utilities, there is a langs.json file.
In this one, line 2 change:
"langs": ["en", "fr"],
with
"langs": ["en", "fr", "sp"],

and after, in ech group of translation add your line of spanish translation that start with sp:

hoksilato2 commented:
Awesome! But this eddi version will be official? I see that needs to be recompiled with next version.

StivlWrith commented:
It is a pull request, just a proposal with some modifications to bring a use in several languages ​​of this software. So far it has not been adopted by the EDCD team, so we can say that is not an official version.

Tkael commented:
This is not a small change and we have not been able to freeze development long enough to give it the attention it deserves, unfortunately.

hoksilato2 commented:
I have taken the liberty of creating this translation project for both Spanish and German. I hope this is not a problem. Anyone who wants to translate can do so through this link:

https://poeditor.com/join/project/aKqOkmx76j

I adapted the json file to make the translation a bit easier.

Please, if this can cause a problem I will close the project ipso facto.

fbrossar25 commented:
Well I hope EDDI's team will have the time to add this feature in a release someday ^^

richardbuckle commented:
Hopefully after EDDI 3.0 goes final we will be able to take stock and incorporate the IL8N. We haven't forgotten about it.

hoksilato2 commented:
i'm glad to announce that EDDI has been translated to spanish. In the next link you can find a lang.json file with english, french and spanish, and prepared for translate german too.

https://github.com/hoksilato2/lang.json-for-EDDI-i18n

Special thanks to @StivlWrith who create this EDDI mod.

PD: I have found some errors in the French translation, but I have not dared to change them. A review  may be necessary (materials for example).

I hope that soon comes an official version with which you can add this file lang.json.

Thank you all.

richardbuckle commented on Utilities/I18N.cs:
```diff
@@ -0,0 +1,212 @@
+        // contains each langs defined in langs.json
+        private static List<String> availableLangs;
+        // Can't use Utilities.Constants there because of cyclic dependecies
+        private static string dataDir = Environment.GetEnvironmentVariable("AppData") + @"\EDDI\";
```
Let's see if we can find a way to break the cyclic dependency here.

fbrossar25 replied:
The only solution i've found was not to place something localized (or localizable) in Constants.cs file.
A "LocalizedConstants.cs" where constants like 'ENVIRONMENT_WITCH_SPACE' or 'VEHICLE_SHIP' will be placed may help. With this I think we can get rid of that cyclic dependency.

richardbuckle replied:
Don’t worry, as soon as I migrate the data to RESX files the dependency will go away by itself.

richardbuckle commented:
I've had a chance to look at this now. Thanks for a great deal of hard work!

Is there a strong reason we are putting all languages into a single big JSON file that is kept in RAM, rather than Microsoft's way using resource DLLs in satellite assemblies? The latter approach strikes me as having a few advantages:

- less memory pressure because the runtime only loads the satellite assemblies it needs, in file-mapped purgeable RAM, as opposed to keeping a very large dictionary object in non-purgeable RAM all the time
- loading a satellite assembly is much faster than parsing a large JSON file at runtime
- easier workflow and reduced merge conflicts as each translator will be working on their own resource files, decoupled from the others

Thoughts welcomed. Now that we have the data I'm willing to take on the task of re-working it into satellite assemblies if we decide that is the best way to go. Although I'll be away Thursday-Monday.

Documentation links:
https://docs.microsoft.com/en-us/dotnet/standard/globalization-localization/
https://docs.microsoft.com/en-us/dotnet/framework/resources/index
https://docs.microsoft.com/en-us/dotnet/framework/wpf/advanced/wpf-globalization-and-localization-overview

richardbuckle commented:
@hoksilato2 if you think you have found errors but are not sure, please open an issue in @fbrossar25 's repo and I'm sure they will be happy to respond.

Tkael commented:
I suggested (early on) the use of a json file rather than xml.
The idea of using .dll files never occurred to me at the time.

hoksilato2 commented:
French community has knowed for this. their forum it's active :) (im wathing you french people... muahaha :P)

richardbuckle commented:
I have reviewed both the lands.json in this PR and @hoksilato2's one. The French text differs quite heavily. While I can read French, I am not a native speaker, so am not qualified to decide which is better. I shall take obvious typo corrections from @hoksilato2's json but otherwise keep to the JSON in this PR.

Once I have finished everything will be up on OneSky for people to review.

richardbuckle commented:
What is the reason for localisiing `Aquarii` to `vɛʁso` and `Arietis` to `belje` please?

hoksilato2 commented:
Hello!!

I don't know why. I notified to French community for some issues on translate. Me and the others Spanish translators used English strings as reference.

fbrossar25 commented:
Hello !
it looks like a copy & paste mistake, as system's and faction's names are not translated.
Sadly i don't have the time to look in details each entry as I'm preparing my master's degree but I will answer as soon as possible if you have any questions ^^

Beondred commented:
I'd like to make an Italian translation. Would you consider implementing it? :)

richardbuckle commented:
Yes @Beondred that would be very welcome! But please wait just another day or two as I am about the change the data format and put the project on OneSky, which will make the job dramatically easier!

Beondred commented:
I'll be waiting! :D

Beondred commented:
Hi guys: any news?

richardbuckle commented:
Please be patient just a while longer. It is much more work than I expected!

hoksilato2 commented:
We finally have 3.0.0 EDDI version! Now is time to implement i18N, right? :-P

PD: no pressure hehe!

richardbuckle commented:
I am actively working on it and getting very close to releasing it in beta.

richardbuckle commented:
Superseded by #503

richardbuckle commented:
@fbrossar25 @hoksilato2 I would welcome your input on the small number of additional French translations that are needed at https://osxtyfi.oneskyapp.com/admin/project/dashboard/project/145348
